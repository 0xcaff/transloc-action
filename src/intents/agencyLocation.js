// @flow

import type { DialogflowApp } from "actions-on-google";
import { fail } from "../result";
import { findUserAgency, setUserAgency } from "../agencies";
import { nextBusWithAgency } from "./nextBus";

export const agencyLocation = async (app: DialogflowApp): Promise<void> => {
  const deviceLocation = app.getDeviceLocation();
  if (!deviceLocation || !deviceLocation.coordinates) {
    fail(
      app,
      `I couldn't find your current location.`,
      "failed to find current location"
    );

    return;
  }

  const maybeAgency = await findUserAgency(app, deviceLocation);
  if (maybeAgency.type === "DELEGATING") {
    return;
  }

  const agency = maybeAgency.value;
  // TODO: Say Something About Saving
  setUserAgency(app, agency);

  await nextBusWithAgency(app, agency);
};
