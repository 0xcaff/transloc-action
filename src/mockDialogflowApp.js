// @flow
import type {
  Response,
  DeviceLocation,
  SupportedPermission,
  SurfaceCapability,
  SimpleResponse,
  Context
} from "actions-on-google";
import {
  RichResponse,
  List,
  OptionItem
} from "actions-on-google/response-builder";
import type { UserData } from "./userData";

type RecordedPermission = {
  context: string,
  permission: SupportedPermission,
  dialogState?: Object
};

type RecordedResponse =
  | { type: "TELL", response: Response }
  | { type: "ASK_FOR_PERMISSION", permission: RecordedPermission }
  | { type: "ASK_WITH_LIST", list: List };

// A mock DialogflowApplication. It is used to pass in arguments and collect.
export class MockDialogflowApp {
  // Output Field
  response: ?RecordedResponse = null;
  contextOut: Map<string, Context<Object>> = new Map();

  // Input Fields
  args: Map<string, any>;
  permissionGranted: boolean = false;
  deviceLocation: ?DeviceLocation;
  selectedOption: ?string;
  surfaceCapabilities: Set<SurfaceCapability> = new Set();
  context: Map<string, Context<Object>> = new Map();
  userStorage: UserData = {
    agency_id: 643
  };

  constructor(args: Map<string, any> = new Map()) {
    this.args = args;
  }

  tell = (speechResponse: Response): ?Object => {
    if (this.response) {
      throw new TypeError("Multiple responses received.");
    }

    this.response = { type: "TELL", response: speechResponse };
  };

  getArgument = (argName: string): any => this.args.get(argName);

  isPermissionGranted = (): boolean => this.permissionGranted;

  askForPermission = (
    context: string,
    permission: SupportedPermission,
    dialogState?: Object
  ): ?Object => {
    if (this.response) {
      throw new TypeError("Multiple responses received.");
    }

    this.response = {
      type: "ASK_FOR_PERMISSION",
      permission: { context, permission, dialogState }
    };
  };

  askWithList = (
    inputPrompt: string | RichResponse | SimpleResponse,
    list: List
  ): Object => {
    if (this.response) {
      throw new TypeError(`Multiple responses received.`);
    }

    this.response = {
      type: "ASK_WITH_LIST",
      list
    };

    return {};
  };

  buildRichResponse = (otherResponse: ?RichResponse): RichResponse => {
    return new RichResponse(otherResponse);
  };

  buildList = (title: string): List => new List(title);

  buildOptionItem = (
    key: ?string,
    synonyms: string | string[] | void
  ): OptionItem => {
    const optionItem = new OptionItem();
    if (key) {
      optionItem.setKey(key);
    }
    if (synonyms) {
      optionItem.addSynonyms(synonyms);
    }
    return optionItem;
  };

  getDeviceLocation = (): ?DeviceLocation => this.deviceLocation;

  getSelectedOption = (): ?string => this.selectedOption;

  hasSurfaceCapability = (cap: SurfaceCapability): boolean =>
    this.surfaceCapabilities.has(cap);

  getContext = (name: string) => this.context.get(name);

  setContext = (name: string, lifespan: ?number, params: Object) => {
    this.contextOut.set(name, {
      name,
      lifespan: lifespan || 0,
      parameters: params
    });
  };

  SupportedPermissions = {
    NAME: "NAME",
    DEVICE_PRECISE_LOCATION: "DEVICE_PRECISE_LOCATION",
    DEVICE_COARSE_LOCATION: "DEVICE_COARSE_LOCATION",
    UPDATE: "UPDATE"
  };

  SurfaceCapabilities = {
    AUDIO_OUTPUT: "actions.capability.AUDIO_OUTPUT",
    SCREEN_OUTPUT: "actions.capability.SCREEN_OUTPUT"
  };
}
