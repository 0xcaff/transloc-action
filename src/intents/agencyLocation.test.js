// @flow
jest.mock("../now");

import { agencyLocation } from "./agencyLocation";
import type { DeviceLocation } from "actions-on-google";
import { MockDialogflowApp } from "../mockDialogflowApp";

const location: DeviceLocation = {
  coordinates: { latitude: 43.082978, longitude: -77.677036 },
  address: "105 Lomb Memorial Dr",
  zipCode: "14623",
  city: "Rochester, NY"
};

it(`sets the agency based on the location`, async () => {
  const app = new MockDialogflowApp();
  app.deviceLocation = location;
  app.permissionGranted = true;
  app.userStorage = {};

  await agencyLocation((app: any));

  expect(app.response).toMatchSnapshot();
  expect(app.userStorage).toMatchSnapshot();
});
