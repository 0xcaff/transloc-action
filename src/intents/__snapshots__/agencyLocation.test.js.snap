// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets the agency based on the location 1`] = `
Object {
  "response": "The following buses are arriving at Gleason Circle. Perkins - Park Point in 11 minutes; Colony Manor / The Province in 16 minutes; RIT Inn in 16 minutes; Racquet Club in 27 minutes; Perkins - Park Point in 32 minutes.",
  "type": "TELL",
}
`;

exports[`sets the agency based on the location 2`] = `
Object {
  "agency_id": 283,
}
`;
